a10networks/a10-octavia
# [STACK-2356][STACK-2371] a10-octavia retry too long and make sure ctx counters will always decrease

## Description
If Bug Fix:
- Severity Level High
- Issue Description
1. acos-client may takes longer than 300 second for retry.
2. And after acos-client retry, a10-octavia still do another round of retry. And cause the retry takes very very long.
3. In logs from QA, seems counters is not decrease in some case.


acos-client PR: https://github.com/a10networks/acos-client/pull/347

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2356
https://a10networks.atlassian.net/browse/STACK-2371

## Technical Approach
1. change acos-client retry mechanism to use time.time() to calculate actual retry time.
2. in a10-octavia, when acos-client already do the retry then adjust retries correctly.
3. Make sure counters will always decrease
- When exceptions happen in flow notification handler, reset counters.
- prevent ctx counters overflow.
- taskflow maybe stopped unexpectedly, so use try-finally to make sure ctx counter decreasing will always be handled.

## Config Changes
**This is only required if the config has been updated in this PR**
- Required: Provide example config snippet within pre block

<pre>
<b>
[a10_controller_worker]
amp_active_retries = 59
amp_active_wait_sec = 10
amp_vcs_retries = 59
amp_vcs_wait_sec = 10
</b>
</pre>


## Test Cases
- a10-octavia retry time test
1. delete instance when openstack running following tasks VThunderComputeConnectivityWait, ConfigureaVCSBackup, VCSSyncWait and GetMasterVThunder
2. measure the a10-octavia retry time. And the retry time should follow configuration option values.

- workers synchronization test:
1. create/delete 10 loadbalancer at the same time

## Manual Testing
- a10-octavia retry time test
1. rpdb on target tasks
2. delete instance when rpdb stop the process, and then continue it.
3. trace log to see how long it take for retry

- workers synchronization test:
```
openstack loadbalancer create --vip-subnet-id tp91 --name vip1
openstack loadbalancer create --vip-subnet-id tp90 --name vip2
openstack loadbalancer create --vip-subnet-id tp92 --name vip3
openstack loadbalancer create --vip-subnet-id tp91 --name vip4
openstack loadbalancer create --vip-subnet-id tp91 --name vip5
openstack loadbalancer create --vip-subnet-id tp91 --name vip6
openstack loadbalancer create --vip-subnet-id tp91 --name vip7
openstack loadbalancer create --vip-subnet-id tp91 --name vip8
openstack loadbalancer create --vip-subnet-id tp91 --name vip9
openstack loadbalancer create --vip-subnet-id tp91 --name vip10

stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer list
+--------------------------------------+-------+----------------------------------+----------------+---------------------+----------+
| id                                   | name  | project_id                       | vip_address    | provisioning_status | provider |
+--------------------------------------+-------+----------------------------------+----------------+---------------------+----------+
| e7592351-2f0f-4096-9bea-b88730c50c16 | vip1  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.79  | ACTIVE              | a10      |
| 3aa77e66-1b53-4026-a888-385eccf262f6 | vip2  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.90.241 | ACTIVE              | a10      |
| 05604ddc-5894-46a6-ace6-a98c1650401f | vip3  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.92.221 | ACTIVE              | a10      |
| 87cd60f0-c31f-4d9b-b12f-5c8173c1b99b | vip4  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.155 | ACTIVE              | a10      |
| 23a8d885-29ac-4be4-8953-cbe59eed6f28 | vip5  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.163 | ACTIVE              | a10      |
| 9fe32e4b-8d73-4e9f-b5bc-ebefec2df444 | vip6  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.159 | ACTIVE              | a10      |
| bc19acd4-e3fd-4d84-a7e8-15fe840d3fbc | vip7  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.156 | ACTIVE              | a10      |
| bf9840b4-d359-46a4-aa92-6557009293e1 | vip8  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.41  | ACTIVE              | a10      |
| eb4a8796-5818-476c-9ec6-152bd400a2cd | vip9  | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.72  | ACTIVE              | a10      |
| c291b6cb-5d6e-4d7d-8105-6137a0a5bf91 | vip10 | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.136 | ACTIVE              | a10      |
+--------------------------------------+-------+----------------------------------+----------------+---------------------+----------+

stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip1
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip2
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip3
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip4
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip5
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip6
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip7
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip8
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip9
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer delete vip10
stack@openstack-4:~/source/a10-octavia/vThunder$ openstack loadbalancer list

```
